code/game/machinery/computer/camera.dm

```diff
@@ -21,225 +21,226 @@ var/camera_cache_id = 1
+	if(!emagged)
+		emagged = 1
+		user << "\blue You have authorized full network access!"
+		ui_interact(user)
+	else
+		ui_interact(user)
 
-		var/cameras[0]
-		for(var/obj/machinery/camera/C in L)
-			var/cam[0]
-			cam["name"] = C.c_tag
-			cam["deact"] = !C.can_use()
-			cam["camera"] = "\ref[C]"
-			cam["x"] = C.x
-			cam["y"] = C.y
-			cam["z"] = C.z
+/obj/machinery/computer/security/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+	if(src.z > 6) return
+	if(stat & (NOPOWER|BROKEN)) return
+	if(user.stat) return
+
+	var/data[0]
+
+
+	data["current"] = null
+
+	var/list/L = list()
+	for (var/obj/machinery/camera/C in cameranet.cameras)
+		if(can_access_camera(C))
+			L.Add(C)
+
+	cameranet.process_sort()
+
+	var/cameras[0]
+	for(var/obj/machinery/camera/C in L)
+		var/cam[0]
+		cam["name"] = C.c_tag
+		cam["deact"] = !C.can_use()
+		cam["camera"] = "\ref[C]"
+		cam["x"] = C.x
+		cam["y"] = C.y
+		cam["z"] = C.z
+
+		cameras[++cameras.len] = cam
+
+		if(C == current)
+			data["current"] = cam
+
+	data["cameras"] = cameras
+
+	tempnets.Cut()
+	if(emagged)
+		access = list(access_captain) // Assume captain level access when emagged
+		data["emagged"] = 1
+	if(isAI(user) || isrobot(user))
+		access = list(access_captain) // Assume captain level access when AI
+
+	// Loop through the ID's permission, and check which networks the ID has access to.
+	for(var/l in networks) // Loop through networks.
+		for(var/m in networks[l]) // Loop through access levels of the networks.
+			if(m in access)
+				if(l in network) // Checks if the network is currently active.
+					tempnets.Add(list(list("name" = l, "active" = 1)))
+				else
+					tempnets.Add(list(list("name" = l, "active" = 0)))
+				break
+	if(tempnets.len)
+		data["networks"] = tempnets
 
-			cameras[++cameras.len] = cam
+	ui = nanomanager.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if (!ui)
+		ui = new(user, src, ui_key, "sec_camera.tmpl", "Camera Console", 900, 800)
 
-			if(C == current)
-				data["current"] = cam
+		// adding a template with the key "mapContent" enables the map ui functionality
+		ui.add_template("mapContent", "sec_camera_map_content.tmpl")
+		// adding a template with the key "mapHeader" replaces the map header content
+		ui.add_template("mapHeader", "sec_camera_map_header.tmpl")
 
-		data["cameras"] = cameras
+		ui.set_initial_data(data)
+		ui.open()
+		ui.set_auto_update(1)
 
-		tempnets.Cut()
-		if(emagged)
-			access = list(access_captain) // Assume captain level access when emagged
-			data["emagged"] = 1
-		if(isAI(user) || isrobot(user))
-			access = list(access_captain) // Assume captain level access when AI
+/obj/machinery/computer/security/Topic(href, href_list)
+	if(href_list["switchTo"])
+		if(src.z>6 || stat&(NOPOWER|BROKEN)) return
+		if(usr.stat || ((get_dist(usr, src) > 1 || !( usr.canmove ) || usr.blinded) && !istype(usr, /mob/living/silicon))) return

```

Comment:
@Fox-McCloud I don't know why you hate the helpers. It's almost always better to use an available helper proc.

Especially when, say, you could want to path drones off silicon, but not have them pass issilicon. For example.